parsing/parsetree.mli

```diff
@@ -506,100 +568,114 @@ and type_extension =
      ptyext_constructors: extension_constructor list;
      ptyext_private: private_flag;
      ptyext_loc: Location.t;
-     ptyext_attributes: attributes;   (* ... [@@id1] [@@id2] *)
+     ptyext_attributes: attributes;  (** ... [\@\@id1] [\@\@id2] *)
     }
-(*
-  type t += ...
+(**
+  Values of type [type_extension] represents [type t += ...].
 *)
 
 and extension_constructor =
     {
      pext_name: string loc;
      pext_kind : extension_constructor_kind;
      pext_loc : Location.t;
-     pext_attributes: attributes; (* C of ... [@id1] [@id2] *)
+     pext_attributes: attributes;  (** [C of ... [\@id1] [\@id2]] *)
    }
 
-(* exception E *)
 and type_exception =
   {
     ptyexn_constructor: extension_constructor;
     ptyexn_loc: Location.t;
-    ptyexn_attributes: attributes; (* ... [@@id1] [@@id2] *)
+    ptyexn_attributes: attributes;  (** [... [\@\@id1] [\@\@id2]] *)
   }
+(** Values of type [type_exception] represents [exception E] *)
 
 and extension_constructor_kind =
-    Pext_decl of string loc list * constructor_arguments * core_type option
-      (*
-         | C of T1 * ... * Tn     ([], [T1; ...; Tn], None)
-         | C: T0                  ([], [], Some T0)
-         | C: T1 * ... * Tn -> T0 ([], [T1; ...; Tn], Some T0)
-         | C: 'a... . T1... -> T0 (['a;...]; [T1;...], Some T0)
+  | Pext_decl of string loc list * constructor_arguments * core_type option
+      (** [Pext_decl(l, c_args, t_opt)]
+          represents the new construtor description of:
+          - [C of T1 * ... * Tn] when:
+               {ul {- [l] is [[]],}
+                   {- [c_args] is [[T1; ...; Tn]],}
+                   {- [t_opt] is [None]}.}
+          - [C: T0] when
+               {ul {- [l] is [[]],}
+                   {- [c_args] is [[]],}
+                   {- [t_opt] is [Some T0].}}
+          - [C: T1 * ... * Tn -> T0] when
+               {ul {- [l] is [[]],}
+                   {- [c_args] is [[T1; ...; Tn]],}
+                   {- [t_opt] is [Some T0].}}
+          - [C: 'a... . T1 * ... * Tn -> T0] when
+               {ul {- [l] is [['a;...]],}
+                   {- [c_args] is [[T1; ... ; Tn]],}
+                   {- [t_opt] is [Some T0].}}
```

Comment:
Taking this case, as an illustration for the pattern mattrix presentation, it would become:
```
         - [C of T1 * ... * Tn] when [l], [c_args], [t_opt] is [[]], [[T1; ...; Tn]], [None]
         - [C: T0] when [l], [c_args], [t_opt] is  [[]], [[]], [Some T0]
         - [C: T1 * ... * Tn -> T0] when [l], [c_args], [t_opt] is [[]], [[T1; ...; Tn]],  [Some T0]
         - [C: 'a... . T1 * ... * Tn -> T0] when [l], [c_args], [t_opt]  is  [['a;...]],  [[T1; ... ; Tn], [Some T0]
```
another possibility might be to use the named tuple variant used in some other cases:
```
         - [C of T1 * ... * Tn] when [l]=[[]], c_args=[[T1; ...; Tn]], t_opt=[None]
         - [C: T0] when  [l]=[[]], [c_args]=[[]], [t_opt]=[Some T0]
         - [C: T1 * ... * Tn -> T0] when  [l]=[[]], [c_args]=[[T1; ...; Tn]],  [Some T0]
         - [C: 'a... . T1 * ... * Tn -> T0] when [l]=[['a;...]],  [c_args]=[[T1; ... ; Tn], [t_opt]=[Some T0]
```
which may have the advantage of making the column of the pattern matrix more visible.

Reply:
```
| Pext_decl(vars, args, ret)    | vars         | args         | ret    |
| ----------------------------- | ------------ | ------------ | ------ |
| C of T1 * .. * Tn             | []           | [T1; ..; Tn] | None   |
| C : T1 * .. * Tn -> T         | []           | [T1; ..; Tn] | Some T |
| C : 'a .. 'z. T1 * .. Tn -> T | ['a; ..; 'z] | [T1; ..; Tn] | Some T |
```

Reply:
I agree, and it is yet another argument to have tables in odoc (and update to fully using odoc).

Reply:
Slightly more:

1. I didn't find your own proposals very readable compared to what's already in the PR, but it may be because you didn't break lines whereas you would actually keep the sub-list design?

2. Do we really want to index by example, and for each example detail all components of the definition? Personally I find myself wondering more often "what does this specific argument mean", and then having two representative examples where it is used would be enough. This would suggest something like:

> - `vars : string loc list` is a list of existential GADT variables:
>   + `K of T` or `K : T1 * T2 -> T` have `[]` 
>   + `K : 'a 'b 'c . T1 * T2 -> T` has `['a; 'b; 'c]`
> - `c_args : constructor_arguments` is the list of constructor arguments
>   + `K of T1 * T2 * T3` and `K : T1 * T2 * T3 -> T` have `[T1; T2; T3]`
> - `t_opt` is the return type of a GADT constructor:
>    + `K of T1 * T2` has `None`
>    + `K : T1 * T2 -> T` has `Some T`

... now that I did the effort of typing it all up, I'm not so convinced by the result, it makes sense for certain ways to use the types but it trades of form of redundancy (repeating argument names several times) for another (repeating concrete examples several times). The two-dimensional array looks best.

Reply:
I agree that none of these format are completely satisfying, but indeed a consistent format that works for all cases seems really difficult to find...
The problem with verbatim environment is that it is (AFAIK) it is not possible to include references in verbatim environment, and I find these references really handy in such big types definition...
But once tables are included in odoc, I'll gladly use them in this file! (if not having them in ocamldoc is ok)

Reply:
It is also might be worthwhile feature request to have links in verbatim environment. On the matter of readability, one issue is that readability in the mli file does not translate to readability in the html format.

Reply:
> It is also might be worthwhile feature request to have links in verbatim environment

I totally agree! It can also be useful in code environment when giving examples.